pkg/apis/apps/v1alpha1/staticpod_types.go

```diff
@@ -0,0 +1,145 @@
+	Type StaticPodUpgradeStrategyType `json:"type,omitempty"`
+
+	// Auto upgrade config params. Present only if type = "auto".
+	//+optional
+	MaxUnavailable *intstr.IntOrString `json:"maxUnavailable,omitempty"`
+}
+
+// StaticPodUpgradeStrategyType is a strategy according to which a static pod gets upgraded.
+type StaticPodUpgradeStrategyType string
+
+const (
+	AutoStaticPodUpgradeStrategyType StaticPodUpgradeStrategyType = "auto"
+	OTAStaticPodUpgradeStrategyType  StaticPodUpgradeStrategyType = "ota"
+)
+
+// StaticPodSpec defines the desired state of StaticPod
+type StaticPodSpec struct {
+	// StaticPodName indicates the static pod desired to be upgraded.
+	StaticPodName string `json:"staticPodName"`
+
+	// StaticPodManifest indicates the Static Pod desired to be upgraded. The corresponding
+	// manifest file name is `StaticPodManifest.yaml`.
+	// +optional
+	StaticPodManifest string `json:"staticPodManifest,omitempty"`
+
+	// Namespace indicates the namespace of target static pod
+	// +optional
+	StaticPodNamespace string `json:"staticPodNamespace,omitempty"`
+
+	// An upgrade strategy to replace existing static pods with new ones.
+	UpgradeStrategy StaticPodUpgradeStrategy `json:"upgradeStrategy,omitempty"`
+
+	// An object that describes the desired upgrade static pod.
+	// +optional
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +kubebuilder:validation:Schemaless
+	Template corev1.PodTemplateSpec `json:"template,omitempty"`
+}
+
+type StaticPodConditionType string
+
+const (
+	// StaticPodUpgradeSuccess means static pods on all nodes have been upgraded to the latest version
+	StaticPodUpgradeSuccess StaticPodConditionType = "UpgradeSuccess"
+
+	// StaticPodUpgradeExecuting means static pods upgrade task is in progress
+	StaticPodUpgradeExecuting StaticPodConditionType = "Upgrading"
+
+	// StaticPodUpgradeFailed means that exist pods failed to upgrade during the upgrade process
+	StaticPodUpgradeFailed StaticPodConditionType = "UpgradeFailed"
+)
+
+// StaticPodCondition describes the state of a StaticPodCondition at a certain point.
+type StaticPodCondition struct {
+	// Type of StaticPod condition.
+	Type StaticPodConditionType `json:"type,omitempty"`
+
+	// Status of the condition, one of True, False, Unknown.
+	Status corev1.ConditionStatus `json:"status,omitempty"`
+
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+
+	// The reason for the condition's last transition.
+	Reason string `json:"reason,omitempty"`
+
+	// A human-readable message indicating details about the transition.
+	Message string `json:"message,omitempty"`
+}
+
+// StaticPodStatus defines the observed state of StaticPod
+type StaticPodStatus struct {
+	// The total number of static pods
+	TotalNumber int32 `json:"totalNumber"`
+
+	// The number of static pods that should be upgraded.
+	DesiredNumber int32 `json:"desiredNumber"`
```

Comment:
I think that DesiredNumber = TotalNumber - UpgradedNumber, so how about rename `DesiredNumber` to `UnupgradedNumber`?

Reply:
Not exactly. For auto mode,
TotalNumber: The total number of nodes that are running the static pod
DesiredNumber: The number of `ready` nodes that are running the static pod

And perhaps we can add a field `UnupgradedNumber` to indicate how many nodes still need be upgraded. But maybe it's a little bit of redundant, since it's straightforward.

Reply:
I think we can define `DesiredNumber` to `ReadyNumber` that used for specifying the number of ready pods.

Reply:
done